rfd/0100-proxy-ssh-grpc.md

```diff
@@ -0,0 +1,199 @@
+that in order to connect to a Node `tsh` must perform at least two SSH handshakes,
+one with the Proxy to setup the connection transport and another with the
+target Node over the transport to establish the user's SSH connection.
+
+## Details
+
+`tsh ssh` needs to connect to the target Node via the Proxy, but it
+doesn't have to use SSH for that communication. A new gRPC service exposed
+by the Proxy could perform the same operations as the existing SSH server
+but without as much overhead required to establish the session. To minimize
+changes both in `tsh` and on Cluster admins, the existing SSH port can be multiplexed
+to accept both SSH and gRPC by leveraging the TLS ALPN protocol `teleport-proxy-grpc-ssh`.
+Any incoming requests on the SSH listener with said ALPN protocol will be routed
+to the gRPC server and all other requests to the SSH server.
+
+Note: a gRPC server is already exposed via the Proxy web address that users the ALPN protocol
+`teleport-proxy-grpc`. In order to not conflict the new ALPN protocol is required. Reusing the
+existing gRPC server is not an option since it is unauthenticated, has aggressive keep alive
+parameters, and is only enabled when TLS Routing is enabled.
+
+### Proto Definition
+
+The specification is modeled after the [ProxyService](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/legacy/client/proto/proxyservice.proto)
+which is a similar transport mechanism leveraged for Proxy Peering.
+
+```proto
+service ConnectionProxyService {
+  // GetClusterDetails provides cluster information from the perspective of the Proxy
+  rpc GetClusterDetails(GetClusterDetailsRequest) returns (GetClusterDetailsResponse);
+  // DialNode establishes a connection to the target host
+  // 
+  // The bidirectional stream is a transport mechanism to send raw data between
+  // client and server. This allows other protocols(SSH) to be sent through the
+  // tunnel.
+  rpc DialNode(stream DialNodeRequest) returns (stream DialNodeResponse);
+}
+
+// Request for DialNode
+//
+// Connection flow
+// -> Target (client)
+// <- Connection (server)
+// <-> Frame/AgentFrame (client/server)
+// 
+// Separating the Frame and AgentFrame allows
+// SSH and the SSH Agent Protocol to be multiplexed
+// over the same connection.
+message DialNodeRequest {
+  oneof payload {
+    // Contains the information to identify and connect
+    // to the target host
+    Target dial_target = 1;
+    // Raw data to transmit
+    DataFrame frame = 2;
+  }
+}
+
+// Response for DialNodeHost
+message DialNodeResponse {
+  oneof payload {
+    // Cluster configuration details
+    ClusterDetails details = 1;
+    // Raw data to transmit
+    DataFrame frame = 2;
+  }
+}
+
+// Protocol represents the supported protocols
+// that can be proxied.
+enum Protocol {
+  PROTOCOL_UNSPECIFIED = 0;
+  // SSH
+  PROTOCOL_SSH = 1;
+  // SSH Agent
+  PROTOCOL_SSH_AGENT = 2;
```

Comment:
Nit: Given how confusing it can be to talk about agent being forwarded to proxy, versus agent being forwarded to the ssh service, could we disambiguate this?  Maybe call it `PROTOCOL_SSH_AGENT_TO_PROXY`, or something like that?  I'd love for the "agent forwarded for the purpose of proxy impersonation of user" to be a bit more explicit.

Reply:
I second Forrest's, specifying just the protocol name can be confusing.

Reply:
I guess what bothers me, is that the two protocols have different destinations/purposes.  One is sideband information to be consumed by the proxy, and one is part of the `client <-> service` connection that just happens to be being forwarded by the proxy.

Looking at this, that's very non-obvious.  I see your point about keeping the enum agnostic so that we can reuse `DataFrame`.  Maybe there's somewhere else we could add some kind of destination discriminant?  Maybe we could use the same `DataFrame`, but extend the top-level enum to have `proxyFrame` and `serviceFrame` variants, instead of a single `frame`?

Idk, might be overthinking it.  I've just found myself disliking how opaque the naming in the proxy/reversetunnel system can be.